File: src/camera_driver/src/camera_node.cpp
#include <ros/ros.h>
#include <sensor_msgs/CompressedImage.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>

int main(int argc, char** argv) {
  ros::init(argc, argv, "camera_node");
  ros::NodeHandle nh;
  ros::NodeHandle pnh("~");

  // --- parameters ---
  int device_id;
  int width, height;
  std::string frame_id;
  bool publish_compressed_only;

  pnh.param<int>("device_id", device_id, 0);
  pnh.param<int>("width", width, 640);
  pnh.param<int>("height", height, 480);
  pnh.param<std::string>("frame_id", frame_id, "camera");
  pnh.param<bool>("compressed_only", publish_compressed_only, false);

  // --- open camera ---
  cv::VideoCapture cap(device_id);
  if (!cap.isOpened()) {
    ROS_ERROR("Failed to open /dev/video%d", device_id);
    return 1;
  }

  cap.set(cv::CAP_PROP_FRAME_WIDTH, width);
  cap.set(cv::CAP_PROP_FRAME_HEIGHT, height);
  // Prefer MJPG for higher effective framerate on USB cameras
  cap.set(cv::CAP_PROP_FOURCC, cv::VideoWriter::fourcc('M', 'J', 'P', 'G'));

  double actual_w = cap.get(cv::CAP_PROP_FRAME_WIDTH);
  double actual_h = cap.get(cv::CAP_PROP_FRAME_HEIGHT);
  ROS_INFO("Camera /dev/video%d opened: %.0fx%.0f", device_id, actual_w, actual_h);

  // --- publishers ---
  image_transport::ImageTransport it(nh);
  image_transport::Publisher pub_raw = it.advertise("camera/image", 1);

  ros::Publisher pub_compressed;
  if (publish_compressed_only) {
    pub_compressed = nh.advertise<sensor_msgs::CompressedImage>("camera/image/compressed", 1);
  }

  cv::Mat frame;
  int seq = 0;

  while (ros::ok()) {
    if (!cap.read(frame) || frame.empty()) {
      ROS_WARN_THROTTLE(2.0, "Camera read failed");
      ros::spinOnce();
      continue;
    }

    ros::Time now = ros::Time::now();
    auto header = std_msgs::Header();
    header.stamp = now;
    header.frame_id = frame_id;
    header.seq = seq++;

    if (publish_compressed_only) {
      // Direct compressed publish (JPEG)
      sensor_msgs::CompressedImage msg;
      msg.header = header;
      msg.format = "jpeg";
      std::vector<uchar> buf;
      cv::imencode(".jpg", frame, buf, {cv::IMWRITE_JPEG_QUALITY, 85});
      msg.data.assign(buf.begin(), buf.end());
      pub_compressed.publish(msg);
    } else {
      // Publish via image_transport (raw + auto-compressed topics)
      sensor_msgs::ImagePtr msg = cv_bridge::CvImage(header, "bgr8", frame).toImageMsg();
      pub_raw.publish(msg);
    }

    ros::spinOnce();
  }

  cap.release();
  return 0;
}

File: src/camera_driver/scripts/camera_node.py
#!/usr/bin/env python3
import rospy
import cv2
from sensor_msgs.msg import CompressedImage

def main():
    rospy.init_node('camera_node', anonymous=False)

    device_id = rospy.get_param('~device_id', 0)
    width     = rospy.get_param('~width', 640)
    height    = rospy.get_param('~height', 480)
    frame_id  = rospy.get_param('~frame_id', 'camera')

    cap = cv2.VideoCapture(device_id, cv2.CAP_V4L2)
    cap.set(cv2.CAP_PROP_FOURCC, cv2.VideoWriter_fourcc(*'MJPG'))
    cap.set(cv2.CAP_PROP_FRAME_WIDTH, width)
    cap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)

    actual_w = cap.get(cv2.CAP_PROP_FRAME_WIDTH)
    actual_h = cap.get(cv2.CAP_PROP_FRAME_HEIGHT)
    if not cap.isOpened():
        rospy.logerr("Failed to open /dev/video%d", device_id)
        return

    rospy.loginfo("Camera /dev/video%d: %.0fx%.0f", device_id, actual_w, actual_h)

    pub = rospy.Publisher('/camera/image/compressed', CompressedImage, queue_size=1)
    seq = 0

    while not rospy.is_shutdown():
        ok, frame = cap.read()
        if not ok or frame is None:
            rospy.logwarn_throttle(2, "Camera read failed")
            rospy.sleep(0.01)
            continue

        ok, jpeg = cv2.imencode('.jpg', frame, [cv2.IMWRITE_JPEG_QUALITY, 85])
        if not ok:
            continue

        msg = CompressedImage()
        msg.header.stamp    = rospy.Time.now()
        msg.header.frame_id = frame_id
        msg.header.seq      = seq
        msg.format          = 'jpeg'
        msg.data            = jpeg.tobytes()

        pub.publish(msg)
        seq += 1

    cap.release()

if __name__ == '__main__':
    main()